runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6327,83 +6327,153 @@ static bool graDepsConflictWithInstanceOfDeps(TR::Node * depNode, TR::Node * nod
+    *
+    * Number Of Slots | Case 1A | Case 1B | Case 2A | Case 2B |
+    *       1         |    8    |   16    |    4    |    8    |
+    *       2         |    16   |   64    |    16   |    32   |
+    *       3         |    32   |   64    |    16   |    32   |
+    *       4         |    64   |   128   |    32   |    64   |
+    *       5         |    64   |   128   |    32   |    64   |
+    *       6         |    64   |   128   |    32   |    64   |
+    *
     */
-   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * sizeofJ9ClassFieldWithinReference) + (sizeofJ9ClassFieldWithinReference * (maxOnsiteCacheSlots != 1) * (cacheCastClass ? 2 : 1));
+   TR::Register *dynamicCacheReg = NULL;
+   
    if (generateDynamicCache)
       {
       TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
-      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
-       * Case 1 : Cast class is runtime variable
-       * Case 2 : Cast class is interface / unresolved
-       * Number Of Slots |  Bytes needed for Case 1 | Bytes needed for Case 2
-       *        1        |              16          |           8
-       *        2        |              64          |           32
-       *        3        |              64          |           32
-       *        4        |              128         |           64
-       *        5        |              128         |           64
-       *        6        |              128         |           64
-       */
       int32_t requestedBytes = 1 << (int) (log2(snippetSizeInBytes-1)+1);
-      traceMsg(comp, "Requested Bytes = %d\n",requestedBytes);
-      // NOTE: For single slot cache, we initialize snippet with addressSize (4/8) assuming which can not be objectClass
-      // In all cases, we use first addressSize bytes to store offset of the circular buffer and rest of buffer will be initialized with 0xf.
+      if (comp->getOption(TR_TraceCG))
+         {
+         traceMsg(comp, "Number Of Dynamic Cache Slots = %d, Caching CastClass: %s\n"
+                        "Bytes needed for Snippet = %d, requested Bytes = %d\n",maxOnsiteCacheSlots, cacheCastClass ? "true" : "false", snippetSizeInBytes, requestedBytes);
+         }
+
       TR_ASSERT_FATAL(maxOnsiteCacheSlots <= 7, "Maximum 7 slots per site allowed because we use a fixed stack allocated buffer to construct the snippet\n");
-      UDATA initialSnippet[16] = { static_cast<UDATA>(addressSize) };
+      U_32 initialSnippet[32] = { 0 };
+
+      if (cacheCastClass)
+         initialSnippet[1] = static_cast<U_32>(sizeofJ9ClassFieldWithinReference * 2);
+      else
+         initialSnippet[isTarget64Bit && !isCompressedRef ? 1 : 0] = static_cast<U_32>(sizeofJ9ClassFieldWithinReference);
+
```

Comment:
This seems needlessly complicated. Why not just store the index at `[0]` always and use `LLGF` always to load and `ST` to store? Then all bytes from offset 4 until `ObjectClassSlot-0` is just alignment bytes.

Reply:
Thanks a lot for the suggestion, made changes in https://github.com/eclipse/openj9/compare/eacc941c50cd958900914c934e54ebeeac8a0e6a..a2430d97387cbde73cf795f75a6d84927a67ee15